dev/PushNotifications/PushNotificationChannel.cpp

```diff
@@ -32,141 +31,89 @@ namespace PushNotificationHelpers
+        {
+            if(PushNotificationHelpers::IsPackagedAppScenario())
+            {
+                auto appUserModelId{ winrt::Microsoft::Helpers::GetAppUserModelId() };
 
-                return m_foregroundHandlers.add(handler);
+                // Register foreground event through the UDK
+                THROW_IF_FAILED(PushNotifications_RegisterNotificationSinkForFullTrustApplication(appUserModelId.get(), this));
             }
-        }
-        else
-        {
-            auto lock = m_lock.lock_exclusive();
-            if (!m_foregroundHandlerCount++)
+            else
             {
                 auto notificationsLongRunningPlatform{ winrt::Microsoft::Helpers::GetNotificationPlatform() };
 
                 wil::unique_cotaskmem_string processName;
                 THROW_IF_FAILED(GetCurrentProcessPath(processName));
 
-                THROW_IF_FAILED(notificationsLongRunningPlatform->RegisterForegroundActivator(this, processName.get()));             
+                // Register foreground event through the LRP
+                THROW_IF_FAILED(notificationsLongRunningPlatform->RegisterForegroundActivator(this, processName.get()));
             }
-            return m_foregroundHandlers.add(handler);
         }
+
+        auto lock { m_lock.lock_exclusive() };
+        return m_foregroundHandlers.add(handler);
     }
 
     void PushNotificationChannel::PushReceived(winrt::event_token const& token) noexcept
     {
-        if (PushNotificationHelpers::IsPackagedAppScenario())
+        bool registeredEvent { false };
         {
-            if (m_channel)
-            {
-                m_channel.PushNotificationReceived(token);
-            }
-            else
-            {
-                auto lock = m_lock.lock_exclusive();
-
-                if (m_foregroundHandlerCount == 1)
-                {
-                    auto appUserModelId{ winrt::Microsoft::Helpers::GetAppUserModelId() };
+            auto lock { m_lock.lock_exclusive() };
+            m_foregroundHandlers.remove(token);
+            registeredEvent = bool(m_foregroundHandlers);
+        }
 
-                    THROW_IF_FAILED(PushNotifications_UnregisterNotificationSinkForFullTrustApplication(appUserModelId.get()));
-                }
+        if(!registeredEvent)
+        {
+            if (PushNotificationHelpers::IsPackagedAppScenario())
+            {
+                auto appUserModelId{ winrt::Microsoft::Helpers::GetAppUserModelId() };
 
-                m_foregroundHandlers.remove(token);
-                --m_foregroundHandlerCount;
+                // Unregister foreground sink in UDK
+                THROW_IF_FAILED(PushNotifications_UnregisterNotificationSinkForFullTrustApplication(appUserModelId.get()));
             }
-        }
-        else
-        {
-            auto lock = m_lock.lock_exclusive();
-            m_foregroundHandlers.remove(token);
-            if (!--m_foregroundHandlerCount)
+            else
```

Comment:
OnRawNotificationReceived method below is broken for Packaged app scenario. I think you will need to call CoCreateInstance again and launch a new process with Payload? Test flow please :)

Reply:
I've tested this on current build and downlevel build. This OnRawNotificationReceived is only for the foreground scenario.

Reply:
Correct. But it won't work for packaged apps the way it is currently implemented right?

Reply:
RegisterForegroundActivator sends the channel object from the packaged app to the LRP. LRP then calls InvokeAll using the channel as a IWpnNotificationSink object.

Reply:
Lets talk offline. I was talking specifically about OnRawNotificationReceived since I cannot add a comment in that block.

Reply:
Resolution: Packaged apps need to be CoCreateInstance'd instead of using ProtocolLaunchHelper.